flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/factory/KubernetesFactoryWithPodTemplateTestBase.java

```diff
@@ -0,0 +1,116 @@
+ * with pod template. These tests will ensure that annotations, labels, imagePullSecrets,
+ * nodeSelector, tolerations, env, init container, sidecar container, volumes from pod template
+ * should be kept after all decorators.
+ */
+public abstract class KubernetesFactoryWithPodTemplateTestBase extends KubernetesTestBase {
+
+    private static final String ENTRY_POINT_CLASS =
+            KubernetesSessionClusterEntrypoint.class.getCanonicalName();
+    private static final int RESOURCE_MEMORY = 1456;
+
+    /**
+     * The result pod variables should be set via {@link #setResultPod()} by implementation class.
+     */
+    protected ObjectMeta resultPodMeta;
+
+    protected PodSpec resultPodSpec;
+
+    @Override
+    protected void setupFlinkConfig() {
+        super.setupFlinkConfig();
+        flinkConfig.set(KubernetesConfigOptionsInternal.ENTRY_POINT_CLASS, ENTRY_POINT_CLASS);
+
+        // Set fixed ports
+        flinkConfig.set(RestOptions.PORT, Constants.REST_PORT);
+        flinkConfig.set(BlobServerOptions.PORT, Integer.toString(Constants.BLOB_SERVER_PORT));
+        flinkConfig.setString(
+                TaskManagerOptions.RPC_PORT, String.valueOf(Constants.TASK_MANAGER_RPC_PORT));
+
+        flinkConfig.set(
+                TaskManagerOptions.TOTAL_PROCESS_MEMORY, MemorySize.ofMebiBytes(RESOURCE_MEMORY));
+    }
+
+    @Override
+    public final void onSetup() throws Exception {
+        setResultPod();
+    }
+
+    protected abstract void setResultPod() throws Exception;
+
+    @Test
+    public void testInitContainerFromPodTemplate() {
+        assertThat(resultPodSpec.getInitContainers().size(), is(1));
+        assertThat(
+                resultPodSpec.getInitContainers().get(0),
+                is(KubernetesPodTemplateTestUtils.getInitContainerInPodTemplate()));
+    }
+
+    @Test
+    public void testSideCarContainerFromPodTemplate() {
+        final Container sideCarContainer =
+                KubernetesPodTemplateTestUtils.getContainerWithName(
+                        resultPodSpec,
+                        KubernetesPodTemplateTestUtils.TESTING_SIDE_CAR_CONTAINER_NAME);
+        assertThat(
+                sideCarContainer,
+                is(KubernetesPodTemplateTestUtils.getSideCarContainerInPodTemplate()));
+    }
+
+    @Test
+    public void testVolumesFromPodTemplate() {
+        assertThat(
+                resultPodSpec.getVolumes(),
+                hasItems(KubernetesPodTemplateTestUtils.getVolumesInPodTemplate()));
+    }
+
+    @Test
+    public void testMainContainerVolumeMountsFromPodTemplate() {
+        final Container mainContainer =
+                KubernetesPodTemplateTestUtils.getContainerWithName(
+                        resultPodSpec, Constants.MAIN_CONTAINER_NAME);
+        assertThat(
+                mainContainer.getVolumeMounts(),
+                hasItems(KubernetesPodTemplateTestUtils.getVolumeMountInPodTemplate()));
+    }
```

Comment:
Looking at the assertions in these tests it is not clear to me why the expected values are what they are. Only when looking at the subclasses of this class one sees that `KubernetesPodTemplateTestUtils.getPodTemplatePath()` is used. This contract is not clear and can easily break. It is much easier to reason about tests if one sees how the test subject is set up and the assertions in one method.

Reply:
I will load the pod template in the base class and only the let the implementation class to decorate it in `getResultPod(FlinkPod podTemplate)`. Then one could know that how the result pod is decorated in the base class.